File: DenoiseStains/cmp_raw_spillover_denoise.m
clearvars;


roi_root_dir = 'E:\LungIMCData\LungROIProcessing\Denoise';
raw_dir = fullfile(roi_root_dir, 'RawROIs');
spillover_dir = fullfile(roi_root_dir, 'SpilloverROIs');
denoise_dir = fullfile(roi_root_dir, 'DenoisedROIs');
cmp_dir = fullfile(roi_root_dir, 'CmpRawSpilloverDenoise');

roi_list = dir(raw_dir);
roi_list = roi_list(3:end);
for sind = 1:length(roi_list)
    roi_name = roi_list(sind).name;
    raw_roi_dir = fullfile(raw_dir, roi_name);
    spillover_roi_dir = fullfile(spillover_dir, roi_name);
    denoise_roi_dir = fullfile(denoise_dir, roi_name);
    cmp_roi_dir = fullfile(cmp_dir, roi_name);
    if ~exist(cmp_roi_dir, 'dir')
        mkdir(cmp_roi_dir)
    end
    
    img_list = dir(raw_roi_dir);
    img_list = img_list(3:end);
    for rind = 1:length(img_list)
        stain_fullname = img_list(rind).name;
        stain_name = stain_fullname(1:end-5);
        raw_roi_path = fullfile(raw_roi_dir, stain_fullname);
        raw_img = imread(raw_roi_path);
        spillover_roi_path = fullfile(spillover_roi_dir, stain_fullname);
        spillover_img = imread(spillover_roi_path);
        denoised_roi_path = fullfile(denoise_roi_dir, stain_fullname);
        denoised_img = imread(denoised_roi_path);
        cmp_roi_path = fullfile(cmp_roi_dir, strcat(stain_name, '.png'));
        high_thresh = quantile(raw_img(raw_img > 0), 0.96);
        montage({raw_img, spillover_img, denoised_img}, 'Size',[1 3], 'DisplayRange', [0, high_thresh]);
        imwrite(getframe(gcf).cdata, cmp_roi_path);
    end
end

File: DenoiseStains/cmp_raw_spillover_denoise.py
# -*- coding: utf-8 -*-

import os, sys
import numpy as np
import argparse, pytz, shutil
from skimage import io
from datetime import datetime


def set_args():
    parser = argparse.ArgumentParser(description = "Comparison of Raw/SpilloverCorrection/Denoising")
    parser.add_argument("--data_root",              type=str,       default="/Data")
    parser.add_argument("--data_type",              type=str,       default="LungROIProcessing")
    parser.add_argument("--denoise_dir",            type=str,       default="Denoise")
    parser.add_argument('--divide_ratio',           type=float,     default=20.0)

    args = parser.parse_args()
    return args


if __name__ == '__main__':
    args = set_args()

    denoise_root_dir = os.path.join(args.data_root, args.data_type, args.denoise_dir)
    raw_dir = os.path.join(denoise_root_dir, "RawROIs")
    spillover_dir = os.path.join(denoise_root_dir, "SpilloverROIs")
    denoise_dir = os.path.join(denoise_root_dir, "DenoisedROIs")
    cmp_dir = os.path.join(denoise_root_dir, "CmpRawSpilloverDenoise")

    roi_lst = sorted([ele for ele in os.listdir(raw_dir) if os.path.isdir(os.path.join(raw_dir, ele))])
    for ind, cur_roi in enumerate(roi_lst):
        if (ind + 1) % 10 == 0:
            print("Processing {}/{}".format(ind+1, len(roi_lst)))
            
        raw_roi_dir = os.path.join(raw_dir, cur_roi)
        spillover_roi_dir = os.path.join(spillover_dir, cur_roi)
        denoise_roi_dir = os.path.join(denoise_dir, cur_roi)
        cmp_roi_dir = os.path.join(cmp_dir, cur_roi)
        if not os.path.exists(cmp_roi_dir):
            os.makedirs(cmp_roi_dir)

        img_lst = [os.path.splitext(ele)[0] for ele in os.listdir(raw_roi_dir) if ele.endswith(".tiff")]
        for cur_img in img_lst:
            raw_img_path = os.path.join(raw_roi_dir, cur_img + ".tiff")
            spillover_img_path = os.path.join(spillover_roi_dir, cur_img + ".tiff")
            denoise_img_path = os.path.join(denoise_roi_dir, cur_img + ".tiff")
            # transform raw
            raw_img = io.imread(raw_img_path, plugin='tifffile').astype(np.float32)
            high_thresh = np.max(raw_img) * 1.0 / args.divide_ratio
            raw_img[raw_img > high_thresh] = high_thresh
            raw_img = (255.0 * (raw_img / high_thresh)).astype(np.uint8)
            # transform spillover
            spillover_img = io.imread(spillover_img_path, plugin='tifffile').astype(np.float32)
            spillover_img[spillover_img > high_thresh] = high_thresh
            spillover_img = (255.0 * (spillover_img / high_thresh)).astype(np.uint8)
            # transform denoise image
            denoise_img = io.imread(denoise_img_path, plugin='tifffile').astype(np.float32)
            denoise_img[denoise_img > high_thresh] = high_thresh
            denoise_img = (255.0 * (denoise_img / high_thresh)).astype(np.uint8)
            # combine three images together
            cmp_img = np.zeros((raw_img.shape[0], raw_img.shape[1] * 3), dtype=np.uint8)
            cmp_img[:,:raw_img.shape[1]] = raw_img
            cmp_img[:,raw_img.shape[1]:2*raw_img.shape[1]] = spillover_img
            cmp_img[:,2*raw_img.shape[1]:] = denoise_img
            # save combined image
            cmp_img_path = os.path.join(cmp_roi_dir, cur_img + ".png")
            io.imsave(cmp_img_path, cmp_img)


